File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement {

    private MapDirection direction;
    private Vector2d position;

    public Animal(Vector2d position) {
        this.direction = MapDirection.NORTH;
        this.position = position;
    }

    @Override
    public String toString() {
        return switch(direction) {
            case NORTH -> "^";
            case SOUTH -> "v";
            case EAST -> ">";
            case WEST -> "<";
        };
    }

    public void move(MoveDirection direction, MoveValidator validator) {
        Vector2d newPosition = position;
        switch (direction) {
            case RIGHT -> this.direction = this.direction.next();
            case LEFT -> this.direction = this.direction.previous();
            case FORWARD -> newPosition = position.add(this.direction.toUnitVector());
            case BACKWARD -> newPosition = this.position.subtract(this.direction.toUnitVector());
        }
        if(newPosition != position && validator.canMoveTo(newPosition)) {
            this.position = newPosition;
        }
    }

    @Override
    public Vector2d getPos() {
        return position;
    }

    public MapDirection getDir() {
        return direction;
    }
}

File: proj-lab2/src/main/kotlin/model/Animal.kt
package model

import util.MapDirection
import util.MoveDirection
import util.Vector2d

class Animal(
    var position: Vector2d,
    private var direction: MapDirection = MapDirection.NORTH,
    private var map: WorldMap
) {
    fun move(direction: MoveDirection) {
        var newPosition: Vector2d = this.position
        when (direction) {
            MoveDirection.RIGHT -> this.direction = this.direction.next()
            MoveDirection.LEFT -> this.direction = this.direction.previous()
            MoveDirection.FORWARD -> newPosition = this.position + this.direction.toUnitVector()
            MoveDirection.BACKWARD -> newPosition = this.position - this.direction.toUnitVector()
        }
        if (newPosition !== this.position && this.map.canMoveTo(newPosition)) {
            this.position = newPosition
        }
    }

    fun isAt(position: Vector2d): Boolean {
        return this.position == position
    }

    override fun toString(): String = this.direction.toString()
}